File: typescript/scripts/comandos/moderação/timeout.ts
//Command timeout
//Exemplo: /timeout <usuário> <tempo> <motivo>
//Linguagem usada: ts
//Author: aquelemesmoojack
//Versão: Discord.JS v14

import { ApplicationCommandType, ApplicationCommandOptionType, PermissionsBitField } from 'discord.js';
import ms from 'ms'

module.exports = {
    name: "timeout",
    description: "Mutar um usuário",
    type: ApplicationCommandType.ChatInput,
    options: [
        {
            name: "membro",
            description: "Mencione o membro que você quer banir",
            type: ApplicationCommandOptionType.User,
            required: true
        },
        {
            name: "tempo",
            description: "Tempo do mute",
            type: ApplicationCommandOptionType.String,
            required: true
        },
        {
            name: "motivo",
            description: "Insira o motivo",
            type: ApplicationCommandOptionType.String,
            required: false
        }
    ],
    run: async (client, interaction) => {
        
        if(interaction.member.permission.has(PermissionsBitField.Flags.ModerateMembers)) {
            const membro = interaction.guild.members.cache.get(interaction.options.getUser("membro").id)
            const motivo = interaction.options.getString("motivo") || "Não informado"
            const tempo = interaction.options.getString("tempo") 

            const tempoMs = ms(tempo)

            interaction.reply({content: "Membro mutado com sucesso", ephemeral: true})
            membro.timeout({reason: [motivo], time: tempoMs})
        } else {
            return interaction.reply({content: "Você não tem permissão", ephemeral: true})
        }
    }
}

File: python/scripts/moderação/timeout.py
# Command timeout
# Exemplo: /timeout <usuário> <tempo> <motivo>
# Linguagem usada: py
# Author: Luis Gomes

import discord
from discord.ext import commands
from discord import app_commands
import datetime

intents = discord.Intents.all()
bot = commands.Bot(command_prefix='!', intents=intents)

@bot.event
async def on_ready():
    print(f'Bot conectado como {bot.user}')
    try:
        synced = await bot.tree.sync()
        print(f'Slash commands sincronizados: {len(synced)} comandos')
    except Exception as e:
        print(f'Erro ao sincronizar slash commands: {e}')

@bot.event
async def on_command_error(ctx, error):
    if isinstance(error, commands.CommandNotFound):
        await ctx.send('Comando não encontrado.')
    elif isinstance(error, commands.MissingPermissions):
        await ctx.send('Você não tem permissão para dar mute membros.')
    elif isinstance(error, commands.BadArgument):
        await ctx.send('Use um ID válido para dar mute um usuário.')
    else:
        await ctx.send(f'Ocorreu um erro: {error}')

@bot.tree.command(name='timeout', description='Mutar um usuario')
@app_commands.describe(usuario='Usuário que será mutado', minutos='Tempo do mute', motivo='Motivo do mute')
@commands.has_permissions(moderate_members=True)
async def timeout(interaction: discord.Interaction, usuario: discord.Member, minutos: int, motivo: str):
    tempo = datetime.timedelta(minutes=minutos)
    try:
        await usuario.edit(timed_out_unitl=discord.utils.utcnow() + tempo, reason=motivo)
        await interaction.response.send_message(f'{usuario.mention} foi colocado em timeout por {minutos} minutos. Motivo: {motivo}')
    except discord.Forbidden:
        await interaction.response.send_message('Não tenho permissão para dar timeout esse usuário.')
    except discord.HTTPException:
        await interaction.response.send_message('Ocorreu um erro ao tentar dar timeout o usuário.')
    except Exception as error:
        await interaction.response.send_message(f'Ocorreu um erro: {error}')

bot.run('TOKEN')
